File: backend/src/main/java/org/certificaciones/funproeibbackend/service/PostulacionService.java
package org.certificaciones.funproeibbackend.service;

import org.certificaciones.funproeibbackend.dto.PostulacionRequest;
import org.certificaciones.funproeibbackend.dto.PostulacionResponse;
import org.certificaciones.funproeibbackend.model.enums.EstadoPostulacion;

import java.util.List;

public interface PostulacionService {
    PostulacionResponse crear(PostulacionRequest request);
    PostulacionResponse obtenerPorId(Long id);
    List<PostulacionResponse> listarPorPrograma(Long idPrograma);
    List<PostulacionResponse> listarPorUsuario(Long idUsuario);
    List<PostulacionResponse> listarPorEstado(EstadoPostulacion estado);
    PostulacionResponse cambiarEstado(Long id, EstadoPostulacion nuevoEstado);
}

File: frontend/src/app/core/services/postulacion.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { PostulacionRequest, PostulacionResponse, EstadoPostulacion } from '../models/postulacion.model';

@Injectable({ providedIn: 'root' })
export class PostulacionService {
  constructor(private api: ApiService) {}

  crear(request: PostulacionRequest): Observable<PostulacionResponse> {
    return this.api.post<PostulacionResponse>('/postulaciones', request);
  }

  obtenerPorId(id: number): Observable<PostulacionResponse> {
    return this.api.get<PostulacionResponse>(`/postulaciones/${id}`);
  }

  listarPorUsuario(idUsuario: number): Observable<PostulacionResponse[]> {
    return this.api.get<PostulacionResponse[]>(`/postulaciones/usuario/${idUsuario}`);
  }

  listarPorPrograma(idPrograma: number): Observable<PostulacionResponse[]> {
    return this.api.get<PostulacionResponse[]>(`/postulaciones/programa/${idPrograma}`);
  }

  cambiarEstado(id: number, estado: EstadoPostulacion): Observable<PostulacionResponse> {
    return this.api.patch<PostulacionResponse>(`/postulaciones/${id}/estado`, { estado });
  }
}
